packages/shared/hooks/useAsync.ts

```diff
@@ -74,88 +74,64 @@ import { useCallback, useState } from 'react';
  *   };
  * });
  */
-export function useAsync<Callback extends (...args: unknown[]) => unknown>(
-  cb?: Callback
+export function useAsync<Args extends unknown[], AttemptData>(
+  cb: (...args: Args) => Promise<AttemptData>
 ) {
-  const [state, setState] =
-    useState<Attempt<Awaited<ReturnType<Callback>>>>(makeEmptyAttempt);
+  const [state, setState] = useState<Attempt<AttemptData>>(makeEmptyAttempt);
+  const isMounted = useIsMounted();
+  const asyncTask = useRef<Promise<AttemptData>>();
 
   const run = useCallback(
-    (
-      ...args: Parameters<Callback> | [AbortSignal, ...Parameters<Callback>]
-    ) => {
-      const [firstArg, ...restOfArgs] = args;
-      let abort = false;
-      let abortSignal: AbortSignal;
-
-      if (firstArg instanceof AbortSignal) {
-        abortSignal = firstArg;
-      }
-
-      if (abortSignal) {
-        abortSignal.addEventListener('abort', () => {
-          abort = true;
-        });
-      }
-
-      return Promise.resolve()
-        .then(() => {
+    (...args: Args) => {
+      setState(prevState => ({
+        ...prevState,
+        status: 'processing',
+      }));
+
+      const promise = cb(...args);
+      asyncTask.current = promise;
+
+      return promise.then(
+        data => {
+          if (!isMounted) {
```

Comment:
@JanKaczmarkiewicz I updated `useAsync` but I'm struggling to get the tests to pass.

During the one marked with `test.only`, I see this:

```
 console.error
    Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
        at TestComponent (/Users/rav/Projects/webapps/node_modules/@testing-library/react-hooks/lib/helpers/createTestHarness.js:22:5)
        at Suspense
        at ErrorBoundary (/Users/rav/Projects/webapps/node_modules/react-error-boundary/dist/react-error-boundary.cjs.js:59:35)

      101 |           }
      102 |
    > 103 |           setState(prevState => ({
          |           ^
      104 |             ...prevState,
      105 |             status: 'success',
      106 |             data,
```

So this kinda indicates that on success, isMounted is still set to true and `useAsync` attempt to `setState`. I'll try to come back to this tomorrow, but I think it has to do with `isMounted` being in deps?

Reply:
You need to call: 
```
if (!isMounted()) {
```
and change the test to be
```
const returnValue = Symbol();
  const { result, unmount } = renderHook(() =>
    useAsync(() => sleep(100).then(() => returnValue))
  );

  let [, run] = result.current;
  let promise: Promise<[symbol, Error]>;
  act(() => {
    promise = run();
  });
  await sleep(50);

  unmount();

  await expect(promise).resolves.toEqual([null, new AbortedSignalError()]);
```
or 
```
let promise: Promise<[unknown, Error]>;
  const TestComponent = () => {
    const asyncFn = useCallback(() => sleep(1000), []);

    const [state, run] = useAsync(asyncFn);

    useEffect(() => {
      promise = run();
    }, [run]);

    return <>{state.status}</>;
  };

  const { unmount } = render(<TestComponent />);

  await screen.findByText('processing');
  unmount();

  await expect(promise).resolves.toEqual([null, new AbortedSignalError()]);
```

Reply:
Thanks a lot, you're a life saver. I was rushing through and I didn't notice that `useIsMounted` returns a function (which makes total sense).